arch/x86/core/pcie.c

```diff
@@ -270,20 +251,23 @@ uint8_t arch_pcie_msi_vectors_allocate(unsigned int priority,
 #endif /* CONFIG_INTEL_VTD_ICTL */
 
 		for (i = 0; i < n_vector; i++) {
-			uint32_t data;
-
-			data = _read_pcie_irq_data(vectors[i].bdf);
-
-			vectors[i].arch.irq = pcie_alloc_irq(vectors[i].bdf);
```

Comment:
This is indeed useless for MSI as it does not make any use of IRQ.
However, due to a design issue on how zephyr manages x86 interrupt vector and irq, you cannot remove that or else line 215 might fail (single msi device) badly if no irq has been allocated to the device(s): irq will be 0.

Zephyr stores vectors on irq basis. It's bad, it should store vector and irq if relevant, in a dedicated data structure. I knew this problem from the start when starting to work on vt-d but no time to change it all (It's quite a big rework actually), so I kept this behavior: it's not generating any harm since as soon as a device is meant to use MSI it will never use the allocated IRQ thus what's being done on IOAPIC will remain dead silent.

Reply:
We should request x86 porting team to fix this issue rather than doing this work around. I can put back this code until Zephyr fix this issue.

Reply:
line 215 will not have any impact on removing this code since this code will come into picture only for multi vector (n_vector > 1). But any way will put back this code with TODO comment to fix the design

Reply:
For 32 vectors this code makes lots of unneeded PCI requests, I was also proposing removing this code.